File: src/main/.net/Model/CartItem.cs
namespace ShoppingApp
{
    public sealed class CartItem
    {
        public string ProductId { get; }

        private int _quantity;
        public int Quantity => _quantity;

        private double _price;
        public double Price => _price;

        public CartItem(string productId, int quantity, double price)
        {
            ProductId = productId;
            _quantity = quantity;
            _price = price;
        }

        public void Add(int quantityToAdd, double priceIncrement)
        {
            if (quantityToAdd <= 0)
                throw new ArgumentException("Must be positive", nameof(quantityToAdd));

            if (priceIncrement <= 0)
                throw new ArgumentException("Must be positive", nameof(priceIncrement));

            _quantity += quantityToAdd;
            _price += priceIncrement;
        }

        public void Remove(int quantityToRemove, double priceDecrement)
        {
            if (quantityToRemove <= 0)
                throw new ArgumentException("Must be positive", nameof(quantityToRemove));

            if (quantityToRemove > _quantity)
                throw new ArgumentException($"Must be less than 'Quantity: {_quantity}'", nameof(quantityToRemove));

            if (priceDecrement <= 0)
                throw new ArgumentException("Must be positive", nameof(priceDecrement));

            if (priceDecrement > _price)
                throw new ArgumentException($"Must be less than 'Price: {_price}'", nameof(priceDecrement));

            _quantity -= quantityToRemove;
            _price -= priceDecrement;
        }
    }
}

File: src/main/java/com/ekart/model/CartItem.java
package com.ekart.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@AllArgsConstructor
public class CartItem {

  private String productId;

  private int quantity;

  private double price;

  public void add(int quantityToAdd, double priceIncrement)
  {
    if (quantityToAdd <= 0)
      throw new IllegalArgumentException("Quantity to add must be positive");

    if (priceIncrement <= 0)
      throw new IllegalArgumentException("Price increment must be positive");

    quantity += quantityToAdd;
    price += priceIncrement;
  }

  public void remove(int quantityToRemove, double priceDecrement)
  {
    if (quantityToRemove <= 0)
      throw new IllegalArgumentException("Quantity to remove Must be positive");

    if (quantityToRemove > quantity)
      throw new IllegalArgumentException("Quantity to remove must be less than or equal to total Quantity");

    if (priceDecrement <= 0)
      throw new IllegalArgumentException("Price decrement must be positive");

    if (priceDecrement > price)
      throw new IllegalArgumentException("Price decrement can not be more than original price");

    quantity -= quantityToRemove;
    price -= priceDecrement;
  }


}
